torch/csrc/jit/passes/onnx/scalar_type_analysis.cpp

```diff
@@ -271,6 +287,45 @@ static void UpdateScalarTypeForOutput(
       CreateProfiledTensorTypeWithScalarType(output_tensor_type, scalar_type));
 }
 
+static void RecoverScalarTypeForOutput(
+    Value* out,
+    const c10::ScalarType& scalar_type) {
+  Node* n = out->node();
+  TORCH_INTERNAL_ASSERT(nullptr != n);
+  const int64_t onnx_type = ScalarTypeToONNXType(scalar_type);
+  Node* cast_node = n->owningGraph()->create(onnx::Cast, 1);
+  cast_node->addInput(out);
+  cast_node->i_(attr::to, onnx_type);
+  cast_node->insertAfter(n);
+  out->replaceAllUsesAfterNodeWith(cast_node, cast_node->output());
+}
+
+static void LowPrecisionCastNodeForStandardOps(Node* n, int opset_version) {
+  auto output_scalar_type =
+      n->output()->type()->castRaw<TensorType>()->scalarType().value();
+  for (size_t i = 0; i < n->inputs().size(); ++i) {
+    auto input_tensor_type =
+        n->input(i)->type()->cast<TensorType>()->scalarType().value();
+    if (input_tensor_type == output_scalar_type) {
+      break;
+    }
+    TORCH_INTERNAL_ASSERT(
+        i == n->inputs().size() - 1,
+        "output node type should at least equal to one of input");
```

Comment:
Why do you expect to see different input types? Shouldn't they be handled in `ImplicitCastForONNX` that is ran before this pass?